File: Solutions/Cpp/roman_to_integer.cpp
/*
    [EASY]
    13. Roman to Integer

    Concepts:
    - string
    - hash table

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 13.34 MB  [Beats 35.91%]
*/

class Solution {
public:
    int romanToInt(string s) {
        int n = s.size();
        int num = 0, i = 0;
        std::map<char, int> r_to_i = {
            {'I', 1},
            {'V', 5},
            {'X', 10},
            {'L', 50},
            {'C', 100},
            {'D', 500},
            {'M', 1000}
        };

        while (i < n) {
            if (i < n-1 && r_to_i[s.at(i)] < r_to_i[s.at(i+1)]) {
                num += r_to_i[s.at(i+1)] - r_to_i[s.at(i)];
                i++;
            } else {
                num += r_to_i[s.at(i)];
            }
            i++;
        }
        return num;
    }
};

File: Solutions/Rust/roman_to_integer.rs
/*
    [EASY]
    13. Roman to Integer

    Concepts:
    - string
    - hash table

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 2.09 MB   [Beats 94.35%]
*/
use std::collections::HashMap;

impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let n = s.len();
        let [mut num, mut i]: [usize; 2] = [0, 0];
        let [mut cur, mut nxt]: [char; 2] = ['a', 'a'];
        let r_to_i: HashMap<char, usize> = HashMap::from([
            ('I', 1),
            ('V', 5),
            ('X', 10),
            ('L', 50),
            ('C', 100),
            ('D', 500),
            ('M', 1000)
        ]);

        while i < n {
            cur = s.chars().nth(i).unwrap();
            if let nxt = s.chars().nth(i+1)
                && nxt.is_some()
                && (r_to_i[&cur] < r_to_i[&nxt.unwrap()])
            {
                num += r_to_i[&nxt.unwrap()] - r_to_i[&cur];
                i += 1;
            } else {
                num += r_to_i[&cur];
            }
            i += 1;
        }

        return num as i32;
    }
}